src/dump.c

```diff
@@ -2291,29 +2313,74 @@ static void jl_insert_method_instances(jl_array_t *list)
+                jl_typemap_entry_t *entry = jl_typemap_assoc_by_type(mt->defs, &search, /*offs*/0, /*subtype*/0);
+                if (entry) {
+                    jl_value_t *mworld = entry->func.value;
+                    if (jl_is_method(mworld) && mi->def.method != (jl_method_t*)mworld && jl_type_morespecific(((jl_method_t*)mworld)->sig, mi->def.method->sig)) {
+                        // There's still a chance this is valid, if any caller made this via `invoke` and the invoke-signature is still valid
+                        assert(mi->backedges);   // should not be NULL if it's on `list`
+                        jl_value_t *invokeTypes;
+                        jl_method_instance_t *caller;
+                        size_t jins = 0, j0, j = 0, nbe = jl_array_len(mi->backedges);
+                        while (j < nbe) {
+                            j0 = j;
+                            j = get_next_backedge(mi->backedges, j, &invokeTypes, &caller);
+                            if (invokeTypes) {
+                                struct jl_typemap_assoc search = {invokeTypes, world, NULL, 0, ~(size_t)0};
+                                entry = jl_typemap_assoc_by_type(mt->defs, &search, /*offs*/0, /*subtype*/0);
+                                if (entry) {
+                                    jl_value_t *imworld = entry->func.value;
+                                    if (jl_is_method(imworld) && mi->def.method == (jl_method_t*)imworld) {
+                                        // this one is OK
+                                        // in case we deleted some earlier ones, move this earlier
+                                        for (; j0 < j; jins++, j0++) {
+                                            jl_array_ptr_set(mi->backedges, jins, jl_array_ptr_ref(mi->backedges, j0));
+                                        }
+                                        continue;
+                                    }
+                                }
+                            }
+                            invalidate_backedges(&remove_code_instance_from_validation, caller, world, "jl_insert_method_instance");
+                            // The codeinst of this mi haven't yet been removed
+                            jl_code_instance_t *codeinst = caller->cache;
+                            while (codeinst) {
+                                remove_code_instance_from_validation(codeinst);
+                                codeinst = codeinst->next;
+                            }
+                        }
+                        jl_array_del_end(mi->backedges, j - jins);
+                        if (jins == 0) {
+                            m = (jl_method_t*)mworld;
+                            valid = 0;
+                        }
                     }
                 }
             }
+            if (!valid) {
+                // None of the callers were valid, so invalidate `mi` too
```

Comment:
```suggestion
                // One of the callees is invalid, so invalidate `mi` too
```

Reply:
My thinking is this: if `caller1` has an outdated call to `callee`, then `caller1` should be invalidated. But if `caller2` also calls `callee`, by an `invoke` where the dispatch is still valid, then both `callee` and `caller2` continue to be valid.